### PR TITLE
feat: full text search on questions page

### PR DESCRIPTION
**Pull request by Revyy**
Initial implementation on adding full text search for the questions page and need some feedback :)
Also needs some cleanup/tweaking before ready.

Currently this implementation only works when using Postgres as the database as I added a GIN index for the concatenation of the title and content, we can then use the full text search capabilities of Postgres to search for questions based on both the title and content of questions.
However, the basic test I added breaks as this approach is not supported by SQLITE.

I don't know how committed you are to supporting both SQLITE and Postgres going forward but I see some different options.

1. Opt for a simpler solution that is supported by both SQLITE and Postgres, something like `WHERE LOWER(questions.content) LIKE LOWER(%query%)`. (Might not scale that well).
2. Try to do something native for both SQLITE and Postgres (might be very complicated to run separate migrations and detect which database is used)
3. Go with Postgres only implementation, loosing SQLITE support :( (at least for this feature)
4.  Something else entirely?

Thoughts on this? :) 

Also, not used to conventional commits so feel free to let me know if it is not up to par!

### CODE REVIEW

File: plugins/qeta-backend/src/database/DatabaseQetaStore.ts [hunk 1]
```diff
@@ -99,6 +99,16 @@ export class DatabaseQetaStore implements QetaStore {
       query.where('questions.author', '=', options.author);
     }
 
+    if (options.searchQuery) {
+      // Only works for postgres but should scale well when used together with the GIN index created in the migration.
+      query.whereRaw(
+        `to_tsvector('english', questions.title || ' ' || questions.content) @@ websearch_to_tsquery(?)`,
```
Comment (drodil) [inline plugins/qeta-backend/src/database/DatabaseQetaStore.ts:105]: I think this is awesome but we should still support both databases. How about doing this conditionally only for postgres and use the where query for other databases? You can get the database type from `knex.client.config.client` as string

File: plugins/qeta-backend/migrations/20230314-search-index.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+/**
+ * @param {import('knex').Knex} knex
+ */
+exports.up = async function up(knex) {
+  await knex.schema.raw(
+    `CREATE INDEX questions_search_content_index ON questions USING GIN (to_tsvector('english', title || ' ' || content))`,
```
Comment (drodil) [inline plugins/qeta-backend/migrations/20230314-search-index.js:6]: I think this does not work in sqlite? Maybe put this also behind database type check for postgres, see the other comment.

**Comment by drodil** [comment]
Thanks a lot for this, it looks great! I proposed some idea how to make this work best on postgres but still support sqlite, how do you think about that?

**Comment by Revyy** [comment]
Glad you appreciate it!

I was looking for a way to check which database was used programatically with knex but didn't have a lot of time yesterday to dig into it, so thank you for the tip! :)
I'll try it with `knex.client.config.client` and update the PR if it works well.
Should be doable to do something clean.

File: plugins/qeta-backend/migrations/20230314-search-index.js (hunk 2 above)
Comment (Revyy) [reply plugins/qeta-backend/migrations/20230314-search-index.js:6]: Yeah it only works for Postgres. I'll try doing the conditional approach 👍

**Comment by drodil** [comment]
Thanks for the contribution, I will ship this right away so you can get it into use :)

**Comment by drodil** [comment]
Hmm, tried this on our application running on postgres but it cannot find any questions by title or content. Also, the question list flickers with the loading box which is a bit odd looking.

**Comment by Revyy** [comment]
> Hmm, tried this on our application running on postgres but it cannot find any questions by title or content. Also, the question list flickers with the loading box which is a bit odd looking.

The postgres search is kind of keyword based where it removes stopwords and builds up tokens (https://www.postgresql.org/docs/current/textsearch-controls.html). Guess this needs to be explained, or we can try to tweak it.
Right now there is a debounce on 400ms, so when you type in the box it should call the API at most every 400ms. Maybe this needs to be increased? 
Or should we add a search button so it only triggers when you click search?

**Comment by drodil** [comment]
Yeah, I tweaked the search a bit so it can also find partial matches. Let's think about the flickering later, now this is great, thanks once more!

**Comment by Revyy** [comment]
Cool! I'll try out the updated version.

And no problem! Was fun. Might pick something else up down the line if you need the help :) 
We are evaluating this for our organisation and so far it looks really promising!

**Comment by drodil** [comment]
Great to hear you are evaluating this! It's still missing quite few features that would be nice addition. Feel free to browse the open issues or come up something totally new if you want to contribute more!

**Pull request merged**